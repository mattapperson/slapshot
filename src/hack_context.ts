expect.extend({
  __slapshot__hack__context(_received, fn) {
    fn({
      // @ts-ignore
      assertionCalls: this.assertionCalls,
      // @ts-ignore
      currentTestName: this.currentTestName,
      // @ts-ignore
      testPath: this.testPath
    });
    return { pass: true, message: () => "expected the slapshot hack to work" };
  }
});
