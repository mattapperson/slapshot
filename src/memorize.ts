import path from "path";
import fs from "fs";
import { Snapshot } from "./types";
import { runInOnlineMode, shouldUpdateSnapshot } from "./utils";
import { safeSnapshot } from "./safeSnapshot";
import { loadSnaps } from "./loadSnaps";
import { mkdir } from "./mkdir";

type SlapshotDataFormat = { thrownError?: any; results: any };

const testMap: any = {};

function returnValues(value: SlapshotDataFormat) {
  if (value.thrownError) {
    throw value.thrownError;
  } else {
    return value.results;
  }
}

export function memorize<ReturnedData = any>(
  snapshotName: string,
  method: () => Promise<ReturnedData> | ReturnedData,
  { pure = true }: { pure?: boolean } = {}
): Promise<ReturnedData> | ReturnedData {
  let jestContext: any;
  // Hack to get access to jest current test context
  // @ts-ignore
  expect().__slapshot__hack__context(context => (jestContext = context));

  let fullSnapshotName = `${jestContext.currentTestName} - ${snapshotName}`;

  if (!pure) {
    if (!testMap[jestContext.currentTestName]) {
      testMap[jestContext.currentTestName] = 0;
    }
    testMap[jestContext.currentTestName] += 1;

    fullSnapshotName = `${fullSnapshotName} (${
      testMap[jestContext.currentTestName]
    })`;
  }

  const basename = `${path.basename(jestContext.testPath)}.snap`;
  const snapFile = path.resolve(
    jestContext.testPath,
    "..",
    "__memorize_snapshots__",
    basename
  );

  let snapshots: Snapshot = {};

  if (fs.existsSync(snapFile)) {
    snapshots = loadSnaps(snapFile);
  }

  const { results: snap } = snapshots[fullSnapshotName] || ({} as Snapshot);
  if (!shouldUpdateSnapshot() && !runInOnlineMode() && !snap) {
    throw new Error(
      `Missing snapshot
    - Method snapshot name: ${fullSnapshotName}
    - Test file: ${jestContext.testPath}

    ${process.env.SLAPSHOT_RERUN_MESSAGE ||
      "Please re-run Jest with the --updateSnapshot flag AND the env var SLAPSHOT_ONLINE=true"}.`.replace(
        new RegExp("        ", "g"),
        ""
      )
    );
  }

  if (!runInOnlineMode()) {
    return safeSnapshot(snap, false);
  }

  let methodResults: SlapshotDataFormat = { results: null, thrownError: null };
  try {
    methodResults.results = method();
  } catch (e) {
    methodResults.thrownError = e;
  }

  if (methodResults.results instanceof Promise) {
    return Promise.resolve(methodResults.results)
      .catch(error => {
        return resolveData(
          snapFile,
          snap,
          fullSnapshotName,
          {
            thrownError: error,
            results: null
          },
          jestContext,
          snapshots
        );
      })
      .then(methodResults => {
        return resolveData(
          snapFile,
          snap,
          fullSnapshotName,
          {
            results: methodResults
          },
          jestContext,
          snapshots
        );
      });
  }

  return resolveData(
    snapFile,
    snap,
    fullSnapshotName,
    methodResults,
    jestContext,
    snapshots
  );
}

function resolveData(
  snapFile: string,
  snap: any,
  fullSnapshotName: string,
  methodResults: SlapshotDataFormat,
  jestContext: any,
  snapshots: Snapshot
) {
  if (!shouldUpdateSnapshot() && runInOnlineMode()) {
    let snapDataToCompare = snap;
    if (typeof snapDataToCompare === "object") {
      snapDataToCompare = JSON.stringify(snapDataToCompare);
    }

    let methodResultsToCompare: any = methodResults;
    if (typeof methodResultsToCompare === "object") {
      methodResultsToCompare = JSON.stringify(methodResultsToCompare);
    }

    if (snap && methodResultsToCompare !== snapDataToCompare) {
      throw new Error(
        `[Warning] Intigration test result does not match the memorized snap file:
        - Method snapshot name: ${fullSnapshotName}
        - Test file: ${jestContext.testPath}

        ${process.env.SLAPSHOT_RERUN_MESSAGE ||
          "Please re-run Jest with the --updateSnapshot flag AND the env var SLAPSHOT_ONLINE=true"}.`.replace(
          new RegExp("        ", "g"),
          ""
        )
      );
    }

    return returnValues(methodResults);
  }

  snapshots[fullSnapshotName] = safeSnapshot(methodResults);

  const newSnap = Object.keys(snapshots)
    .sort()
    .reduce((acc, key) => {
      return (
        acc +
        `\nexports['${key}'] = ${JSON.stringify(snapshots[key], null, 2)}\n`
      );
    }, "");

  mkdir(snapFile);
  fs.writeFileSync(snapFile, newSnap);
  return returnValues(methodResults);
}
