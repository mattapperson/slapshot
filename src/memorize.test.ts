/* eslint-env jest */
import rimraf from "rimraf";
import { memorize } from "./memorize";
import fs from "fs";
import { performance } from "perf_hooks";
import "./hack_context";

const snapshotDir = `${__dirname}/__memorize_snapshots__`;

const fetchData = () => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve({ foo: "bar" });
    }, 200);
  });
};

const complexReturn = (someValue: any = null) => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(someValue || { foo: "bar", method: () => ({ here: "now" }) });
    }, 200);
  });
};

let mockPromise: any;
let mockThunk: any;

beforeEach(() => {
  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.argv = process.argv.filter(e => e !== "-u");
  process.env.SLAPSHOT_ONLINE = undefined;

  mockPromise = jest.fn(fetchData);
  mockThunk = jest.fn(mockPromise);
  rimraf.sync(snapshotDir);
});

test("calls thunk on first run", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  const data = await memorize("a", mockThunk);
  expect(mockPromise.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("writes a __data_snapshot__ file to disk", async () => {
  expect(fs.existsSync(snapshotDir)).toBe(false);
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  await memorize("b", mockThunk);
  expect(fs.existsSync(snapshotDir)).toBe(true);
});

test("resolves from disk on 2nd hit", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const liveStart = performance.now();
  await memorize("c", mockThunk);
  const liveEnd = performance.now();

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const recordedStart = performance.now();
  const data = await memorize("c", jest.fn());
  const recordedEnd = performance.now();

  expect(liveEnd - liveStart).toBeGreaterThan(190);
  expect(recordedEnd - recordedStart).toBeLessThan(10);
  expect(mockPromise.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("calls run but dont update with SLAPSHOT_ONLINE", async () => {
  const mockWithValue = jest.fn(complexReturn);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockWithValue);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  expect(() => {
    const data: any = memorize("d", () => mockWithValue({ foo: "foo" }));
    expect(mockWithValue.mock.calls.length).toBe(2);
    expect(mockWithValue.mock.calls.length).toBe(2);
    expect(data.foo).toBe("foo");
  }).toThrow();

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data2: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data2.foo).toBe("bar");
});

test("calls run still in unit tests with just updateSnapshot", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockThunk);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data: any = await memorize("d", mockThunk);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(mockThunk.mock.calls.length).toBe(1);
  expect(data.foo).toBe("bar");
});

test("calls run and update with updateSnapshot and SLAPSHOT_ONLINE", async () => {
  const mockWithValue = jest.fn(complexReturn);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("d", mockWithValue);

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  const data: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data.foo).toBe("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data2: any = await memorize("d", () => mockWithValue({ foo: "foo" }));
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(mockWithValue.mock.calls.length).toBe(2);
  expect(data2.foo).toBe("foo");
});

test("works with strings and numbers", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  memorize("c", () => 22);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data = memorize("c", () => {});
  expect(data).toBe(22);
});

test("nested APIs with functions throw an error when called unless mocked manualy", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";
  await memorize("complex", complexReturn);

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  const data: any = await memorize("complex", complexReturn);
  expect(data.foo).toBe("bar");
  expect(typeof data.foo).toBe("string");
  expect(typeof data.method).toBe("function");

  expect(() => {
    data.method();
  }).toThrow();
});

test("Impure memorized methods also add call count to name", async () => {
  const result1 = await memorize(
    "complex",
    () => {
      return 22;
    },
    { pure: false }
  );

  const result2 = await memorize(
    "complex",
    () => {
      return 21;
    },
    { pure: false }
  );

  expect(result1).toBe(22);
  expect(result2).toBe(21);
});

test("thrown errors are replayed", async () => {
  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const mockedCB = jest.fn();
  expect(() => {
    memorize("error", () => {
      throw new Error("foo");
    });
  }).toThrowError("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";
  expect(() => {
    memorize("error", mockedCB);
  }).toThrowError("foo");

  expect(mockedCB).not.toBeCalled();
});

test("thrown errors of a custom type are replayed as the correct instance", async () => {
  class CustomError extends Error {
    constructor(message: any) {
      super(message);
      this.name = "CustomError";
    }
  }

  process.argv.push("--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "true";

  const mockedCB = jest.fn();
  expect(() => {
    memorize("error", () => {
      throw new CustomError("foo");
    });
  }).toThrowError("foo");

  process.argv = process.argv.filter(e => e !== "--updateSnapshot");
  process.env.SLAPSHOT_ONLINE = "false";

  try {
    memorize("error", mockedCB);
  } catch (e) {
    expect(e).toBeInstanceOf(CustomError);
  }
});
