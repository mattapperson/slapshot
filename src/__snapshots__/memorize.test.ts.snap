// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No race conditions - works many times out of sync - c - 1 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 2 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 3 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 4 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 5 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 6 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 7 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 8 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 9 -- 0`] = `{"results":22}`;

exports[`No race conditions - works many times out of sync - c - 10 -- 0`] = `{"results":22}`;

exports[`calls run and update with updateSnapshot and SLAPSHOT_ONLINE - d -- 0`] = `{"results":{"foo":"foo"}}`;

exports[`calls run but dont update with SLAPSHOT_ONLINE - d -- 0`] = `{"results":{"foo":"bar","method":"[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"}}`;

exports[`calls run still in unit tests with just updateSnapshot - d -- 0`] = `{"results":{"foo":"bar"}}`;

exports[`calls thunk on first run - a -- 0`] = `{"results":{"foo":"bar"}}`;

exports[`consoles error of non-matching snap when validateSnapshot is not set - c -- 0`] = `{"results":22,"thrownError":null}`;

exports[`deffers to validateSnapshot to validate snapshot when validateSnapshot is a funtion - c -- 0`] = `{"results":22,"thrownError":null}`;

exports[`does not break toMatchSnapshot - c -- 0`] = `{"results":22,"thrownError":null}`;

exports[`does not break toMatchSnapshot: some dumb snapshot 1`] = `22`;

exports[`nested APIs with functions throw an error when called unless mocked manualy - complex -- 0`] = `{"results":{"foo":"bar","method":"[UNMOCKED-UNSNAPSHOTTED-FUNCTION]"}}`;

exports[`resolves from disk on 2nd hit - c -- 0`] = `{"results":{"foo":"bar"}}`;

exports[`thrown errors are replayed - error -- 0`] = `{"results":null,"thrownError":"{\\"type\\":\\"Error\\",\\"message\\":\\"foo\\",\\"name\\":\\"Error\\",\\"stack\\":\\"Error: foo\\\\n    at memorize_1.memorize (/home/coder/projects/slapshot/src/memorize.test.ts:279:13)\\\\n    at Object.memorize (/home/coder/projects/slapshot/src/memorize.ts:62:29)\\\\n    at expect (/home/coder/projects/slapshot/src/memorize.test.ts:278:5)\\\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\\\n    at /home/coder/projects/slapshot/src/memorize.test.ts:281:6\\\\n    at Generator.next (<anonymous>)\\\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\\\n    at new Promise (<anonymous>)\\\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\\\n    at Object.<anonymous>.test (/home/coder/projects/slapshot/src/memorize.test.ts:272:47)\\\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\\\n    at new Promise (<anonymous>)\\\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\\"}"}`;

exports[`thrown errors are replayed with custom properties - error -- 0`] = `{"results":null,"thrownError":"{\\"type\\":\\"Error\\",\\"message\\":\\"foo\\",\\"name\\":\\"Error\\",\\"stack\\":\\"Error: foo\\\\n    at memorize_1.memorize (/home/coder/projects/slapshot/src/memorize.test.ts:307:13)\\\\n    at Object.memorize (/home/coder/projects/slapshot/src/memorize.ts:62:29)\\\\n    at expect (/home/coder/projects/slapshot/src/memorize.test.ts:306:5)\\\\n    at Object.<anonymous> (/home/coder/projects/slapshot/node_modules/expect/build/toThrowMatchers.js:81:11)\\\\n    at Object.throwingMatcher [as toThrowError] (/home/coder/projects/slapshot/node_modules/expect/build/index.js:342:33)\\\\n    at /home/coder/projects/slapshot/src/memorize.test.ts:309:6\\\\n    at Generator.next (<anonymous>)\\\\n    at /home/coder/projects/slapshot/node_modules/tslib/tslib.js:110:75\\\\n    at new Promise (<anonymous>)\\\\n    at Object.__awaiter (/home/coder/projects/slapshot/node_modules/tslib/tslib.js:106:16)\\\\n    at Object.<anonymous>.test (/home/coder/projects/slapshot/src/memorize.test.ts:292:70)\\\\n    at Object.asyncJestTest (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:102:37)\\\\n    at resolve (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:43:12)\\\\n    at new Promise (<anonymous>)\\\\n    at mapper (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:26:19)\\\\n    at promise.then (/home/coder/projects/slapshot/node_modules/jest-jasmine2/build/queueRunner.js:73:41)\\",\\"customProp\\":\\"bar\\"}"}`;

exports[`throws error of non-matching snap when validateSnapshot is set to true - validateSnapshot -- 0`] = `{"results":22,"thrownError":null}`;

exports[`works with JSON string in memorize name - JSON - {"foo":"dog"} -- 0`] = `{"results":{"foo":"bar"}}`;

exports[`works with array value - c -- 0`] = `{"results":[1,2,3],"thrownError":null}`;

exports[`works with null value - c -- 0`] = `{"results":null,"thrownError":null}`;

exports[`works with strings and numbers - c -- 0`] = `{"results":22,"thrownError":null}`;

exports[`works with undefined value - c -- 0`] = `{"thrownError":null}`;

exports[`writes a __snapshot__ file to disk - b -- 0`] = `{"results":{"foo":"bar"}}`;
